httpcore/_async/connection_pool.py

```diff
@@ -238,24 +238,27 @@ def _assign_requests_to_connections(self) -> List[AsyncConnectionInterface]:
         those connections to be handled seperately.
         """
         closing_connections = []
+        idling_connections = {c for c in self._connections if c.is_idle()}
```

Comment:
It seems like we use this collection just for tracking the count of idle connections. Maybe it should just be an integer for simplicity?

Reply:
With integer we would need to recheck all the connections again. We can not just decrement below in the if-elif branches as we could end up going negative etc

Reply:
Heh sorry, you are right. We can just check eg the expired one if it was also idle one and decrement!

Reply:
Does that part significantly change the performance? Did you run any tests? Maybe we are gaining the same performance boost by only using polling?

Reply:
When I checked with `pyinstrument` it showed this exact spot as a hot spot spending time iterating the connections in outer and inner loops. Ill rerun the benchmark to check it again

Reply:
Anyways I pushed now the integer fix

Reply:
Without the fix in `_assign_requests_to_connections` the performance starts to deteriorate when the amount of connections in the pool increases (which is not surprising as there is the loop in the loop).

With 20 concurrency (sync):
![sync_20](https://github.com/encode/httpcore/assets/12939780/e1171d90-f92e-4071-8625-c7a745a7e0db)

With 40 concurrency (sync):
![sync_40](https://github.com/encode/httpcore/assets/12939780/d6315227-31e7-42aa-8489-7ffe22147113)